### PR TITLE
Generate signature for Attributes.

### PR DESCRIPTION
**Pull request by chowchow316**
Signature will be used for cache.

### CODE REVIEW

File: src/signature.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+#define MIXER_CLIENT_SIGNATURE_H_
+
+#include <string>
+#include "include/client.h"
+
+namespace istio {
+namespace mixer_client {
+
+// Generates signature for Attributes.
+std::string GenerateAttributesSignature(const Attributes& attributes);
```
Comment (qiwzhang) [inline src/signature.h]: Just call it GenerateSignature

File: src/signature.cc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        hasher.Update(&attribute.second.value.double_v,
+                      sizeof(attribute.second.value.double_v));
+      case Attributes::Value::ValueType::BOOL:
+        hasher.Update(&attribute.second.value.bool_v,
+                      sizeof(attribute.second.value.bool_v));
+      case Attributes::Value::ValueType::TIME:
+        hasher.Update(&attribute.second.value.time_v,
+                      sizeof(attribute.second.value.time_v));
+    }
+    hasher.Update(kDelimiter, kDelimiterLength);
```
Comment (qiwzhang) [inline src/signature.cc]: need to hash name too; attribute.first.  Put a delimiter between name and value

File: BUILD [hunk 3]
```diff
@@ -68,4 +72,15 @@ cc_test(
         ":simple_lru_cache",
         "//external:googletest_main",
     ],
-)
\ No newline at end of file
+)
+
+cc_test(
+    name = "md5_test",
```
Comment (qiwzhang) [inline BUILD]: need a simple test for GenerateSignature function with all data types.

File: src/signature.cc [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace istio {
+namespace mixer_client {
+namespace {
+const char kDelimiter[] = "\0";
+const int kDelimiterLength = 1;
+}  // namespace
+
+string GenerateAttributesSignature(const Attributes& attributes) {
+  MD5 hasher;
```
Comment (qiwzhang) [inline src/signature.cc]: You may have to wait.  It seem that std::string could not be inside union.

Comment (chowchow316) [inline src/signature.cc]: It seems that std::string can be inside union since C++11: http://en.cppreference.com/w/cpp/language/union

Comment (lizan) [inline src/signature.cc]: std::string can be inside union, but it needs a manually allocation (in-place new) and deallocation to call constructor and destructor of std::string. There are similar codes in Envoy.

**Comment by qiwzhang** [comment]
std::hash doesn't do incremental update.  You have to copy all fields into a big buffer first, then calculate hash.

**Comment by lizan** [comment]
Sorry the previous comment wasn't very clear. I meant specializing std::hash by our implementation, i.e. define `std::hash<Attributes>`. You can put the incremental updater inside it, just like what we do here in `GenerateAttributesSignature`. See how to inject it here: http://en.cppreference.com/w/cpp/utility/hash 

The implementation is a different topic. How big do you expect Attributes could be?

**Comment by qiwzhang** [comment]
I see.  

Then the tricky part is how to implement it.   One way is to hash each name as string, and value as standard type, and combine their hash results.   we may have 20 or more name value pairs,  you need to combine so many hash results, I am afraid that it will loss its randomness.    It will cause signature collision.  It seems there is not good way to combine hash (http://stackoverflow.com/questions/5889238/why-is-xor-the-default-way-to-combine-hashes)

Another way is to use md5 in the implementation of std::hash<Attributes>; incrementally update the origin data, then use md5 to get hash. But its hash result size is size_t.   It is much smaller than MD5 length which is 128 bits.   So I think we better go with current implementation of MD5.

File: src/signature.cc [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  for (const auto& attribute : attributes.attributes) {
+    switch (attribute.second.type) {
+      case Attributes::Value::ValueType::STRING:
+        hasher.Update(attribute.second.value.str_v);
```
Comment (qiwzhang) [inline src/signature.cc]: need break;

File: src/signature_test.cc [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+using std::string;
+using ::google::protobuf::TextFormat;
+
+namespace istio {
+namespace mixer_client {
+namespace {
+
+class SignatureUtilTest : public ::testing::Test {
+ protected:
+  SignatureUtilTest() { attributes_map_.attributes.clear(); }
```
Comment (qiwzhang) [inline src/signature_test.cc]: not need to call clear().

File: src/signature_test.cc [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  EXPECT_EQ("e9c0e5f10e2110ca0636d421da4da29e",
+            MD5::DebugString(GenerateSignature(attributes_map_)));
+
+  AddInt64Pair("int-key", 35);
+  EXPECT_EQ("02da0af21811309e2290c37a703082d5",
+            MD5::DebugString(GenerateSignature(attributes_map_)));
+
+  AddBoolPair("bool-key", true);
+  EXPECT_EQ("b1ffa4e01c3882938d6fd75a575f259a",
+            MD5::DebugString(GenerateSignature(attributes_map_)));
```
Comment (qiwzhang) [inline src/signature_test.cc]: add Time?

File: src/signature_test.cc (hunk 6 above)
Comment (chowchow316) [inline src/signature_test.cc]: Done.

File: src/signature_test.cc (hunk 7 above)
Comment (chowchow316) [inline src/signature_test.cc]: Done.

**Pull request merged**